torch/nn/parameter.py

```diff
@@ -23,8 +36,18 @@ class Parameter(torch.Tensor):
     def __new__(cls, data=None, requires_grad=True):
         if data is None:
             data = torch.empty(0)
-        return torch.Tensor._make_subclass(cls, data, requires_grad)
-
+        if type(data) is torch.Tensor:
+            # For ease of BC maintenance, keep this path for standard Tensor.
+            # Eventually (tm), we should change the behavior for standard Tensor to match.
+            return torch.Tensor._make_subclass(cls, data, requires_grad)
+
+        # Path for custom tensors: set a flag on the instance to indicate parameter-ness.
+        t = data.detach().requires_grad_(requires_grad)
```

Comment:
This is failing parametrization tests, since `remove_parametrizations(..., leave_parametrized=True)` calls `subclass_param.set_(other_subclass_instance)`:
```
RuntimeError: set_storage is not allowed on a Tensor created from .data or .detach().
If your intent is to change the metadata of a Tensor (such as sizes / strides / storage / storage_offset)
without autograd tracking the change, remove the .data / .detach() call and wrap the change in a `with torch.no_grad():` block.
For example, change:
    x.data.set_(y)
to:
    with torch.no_grad():
        x.set_(y)
```

For plain tensors, it looks like `_make_subclass()` creates a new tensor, detaches, then moves the data into it. The line `data.unsafeGetTensorImpl()->set_allow_tensor_metadata_change(true)` seems to help avoid the error for the plain tensor case.
https://github.com/pytorch/pytorch/blob/905efa82ff698f252a44fac83efaa72fe5678f52/torch/csrc/autograd/python_variable.cpp#L380-L396

Is it a bad idea to use `clone()` here instead of `detach()`? @albanD @ezyang

Reply:
I guess we can cheat with this: `torch.Tensor._make_subclass(type(data), requires_grad)` to get exactly the old behavior.

Reply:
@ezyang given that this error was introduced to avoid silent BC-breaking failures when Variable and Tensor were merged. Maybe it would be ok to just remove that whole `allow_tensor_metadata_change` logic?

Reply:
I tried this workaround already but couldn't get it to work:
```python
...
t = torch.Tensor._make_subclass(type(data), data, requires_grad)
t._is_param = True
return t
```
```
RuntimeError: Creating a new Tensor subclass LoggingTensor but the raw Tensor object is already associated to a python object of type LoggingTensor
```

Reply:
OK I changed my mind, using detach() is better and we should do something about allow_tensor_metadata_change. Not sure if removal is the right thing.